libazureinit/src/config.rs

```diff
@@ -0,0 +1,899 @@
+pub struct Wireserver {
+    /// Timeout in seconds for establishing a connection to the wire server.
+    pub connection_timeout_secs: f64,
+
+    /// Timeout in seconds for reading data from the wire server.
+    pub read_timeout_secs: f64,
+
+    /// Total retry timeout in seconds for wire server requests.
+    pub total_retry_timeout_secs: f64,
+}
+
+impl Default for Wireserver {
+    fn default() -> Self {
+        Self {
+            connection_timeout_secs: 2.0,
+            read_timeout_secs: 60.0,
+            total_retry_timeout_secs: 1200.0,
+        }
+    }
+}
+
+/// Telemetry configuration struct.
+///
+/// Configures telemetry behavior, including diagnostic settings.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct Telemetry {
+    /// Flag to enable or disable KVP diagnostics. Defaults to `true`.
+    pub kvp_diagnostics: bool,
+}
+
+impl Default for Telemetry {
+    fn default() -> Self {
+        Self {
+            kvp_diagnostics: true,
+        }
+    }
+}
+
+/// General configuration struct for azure-init.
+///
+/// Aggregates all configuration settings for managing SSH, provisioning, IMDS, media,
+/// and telemetry, supporting loading from file or directory and merging configurations.
+#[derive(Default, Serialize, Deserialize, Debug, Clone)]
+#[serde(default)]
+pub struct Config {
+    pub ssh: Ssh,
+    pub hostname_provisioners: HostnameProvisioners,
+    pub user_provisioners: UserProvisioners,
+    pub password_provisioners: PasswordProvisioners,
+    pub imds: Imds,
+    pub provisioning_media: ProvisioningMedia,
+    pub azure_proxy_agent: AzureProxyAgent,
+    pub wireserver: Wireserver,
+    pub telemetry: Telemetry,
+}
+
+/// Implements `Display` for `Config`, formatting it as a readable TOML string.
+///
+/// Uses `toml::to_string_pretty` to serialize the configuration. If serialization fails,
+/// a fallback message is displayed..
+impl fmt::Display for Config {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            toml::to_string_pretty(self)
+                .unwrap_or_else(|_| "Unable to serialize config.".to_string())
+        )
+    }
+}
+
+/// Loads the configuration for `azure-init`.
+///
+/// This method uses the `Figment` library to load configuration from the following sources,
+/// in order of priority:
+///
+/// 1. **Defaults**: Base configuration from `Config::default()`.
+/// 2. **Main File**: `azure-init.toml`, if present.
+/// 3. **Directory Files**: `.toml` files in `azure-init.d`, sorted lexicographically.
+/// 4. **CLI Overrides**: A file or directory specified via the CLI.
+///
+/// Later sources override earlier ones in case of conflicts.
+impl Config {
+    const BASE_CONFIG: &'static str = "/etc/azure-init.toml";
+    const DROP_IN_CONFIG: &'static str = "/etc/azure-init.d/";
+
+    /// Public wrapper function that uses default locations
+    pub fn load(path: Option<PathBuf>) -> Result<Config, Error> {
```

Comment:
The `///` renders it as documentation and since the fact that this is a wrapper is an internal implementation detail, we shouldn't call it out in public docs. Something like

```Rust
/// Load provisioning configuration.
///
/// In addition to the provided path, configuration will also be loaded from the default locations.
```

Let's developers know it's what they want to use to load config from the filesystem.